django/thunderstore/core/tests/test_utils.py

```diff
@@ -110,3 +111,67 @@ def test_capture_exception_always_raise(settings: Any):
     settings.ALWAYS_RAISE_EXCEPTIONS = True
     with pytest.raises(Exception, match="test"):
         capture_exception(Exception("test"))
+
+
+@pytest.mark.parametrize(
+    ("url", "preferred", "expected"),
+    (
+        (
+            "https://gcdn.thunderstore.io/healthz",
+            None,
+            "https://gcdn.thunderstore.io/healthz",
+        ),
+        (
+            "https://gcdn.thunderstore.io/healthz",
+            "",
+            "https://gcdn.thunderstore.io/healthz",
+        ),
+        (
+            "https://gcdn.thunderstore.io/healthz",
+            "disallowed.thunderstore.io",
+            "https://gcdn.thunderstore.io/healthz",
+        ),
+        (
+            "https://gcdn.thunderstore.io/healthz",
+            "gcdn.thunderstore.io",
+            "https://gcdn.thunderstore.io/healthz",
+        ),
+        (
+            "https://gcdn.thunderstore.io/healthz",
+            "hcdn-1.hcdn.thunderstore.io",
+            "https://hcdn-1.hcdn.thunderstore.io/healthz",
+        ),
+        (
+            "https://gcdn.thunderstore.io",
+            "hcdn-1.hcdn.thunderstore.io",
+            "https://hcdn-1.hcdn.thunderstore.io",
+        ),
+        (
+            "https://hcdn-1.hcdn.thunderstore.io/healthz",
+            "gcdn.thunderstore.io",
+            "https://gcdn.thunderstore.io/healthz",
+        ),
+        (
+            "http://localhost.thunderstore/healthz",
+            "hcdn-1.hcdn.thunderstore.io",
+            "http://hcdn-1.hcdn.thunderstore.io/healthz",
+        ),
+        (
+            "http://localhost:9000/healthz",
+            "hcdn-1.hcdn.thunderstore.io",
+            "http://hcdn-1.hcdn.thunderstore.io/healthz",
+        ),
+        (
+            "/relative/path",
+            "hcdn-1.hcdn.thunderstore.io",
+            "//hcdn-1.hcdn.thunderstore.io/relative/path",
```

Comment:
What's up with the double slash prefix? Is it because we can't know the protocol in this scenario?

It might be fine but it also feels like scenario where we might want to either default to `settings.PROTOCOL` or raise an exception. Exception is probably preferred if this "should never happen". Thoughts?

Reply:
This test was here just to document that the implementation, which relies on urllib method, will return protocol-relative URLs if relative URLs are used.

Changed the implementation to raise instead. While this prevents using the function for some valid cases, I see it unlikely that we'll have those valid cases, so might raise as well. Restoring the support is easy if the need ever raises.